#A/B Test Analysis: Website Conversion Optimization

# Stage 1: Data Environment Setup and Initial Ingestion

### Description
The first phase of the project involves setting up the analytical environment in Google Colab and importing the dataset. This stage ensures that the Python environment has access to the data stored on Google Drive and validates that the file structure is correct for subsequent statistical analysis.

### Key Technical Steps
* **Environment Connection:** Mounting Google Drive and importing the `pandas` and `os` libraries to manage files and dataframes.
* **Data Loading:** Using `pd.read_csv()` to load the A/B test results into a DataFrame named `df`.
* **Structural Inspection:** Executing `df.head()` to preview the first rows and `df.info()` to check for data types, non-null counts, and memory usage.

### Findings
* **Dataset Confirmation:** The file `ab_test_results.csv` was successfully loaded.
* **Data Integrity:** The initial inspection shows that the dataset contains key columns: `ga_session_id` (unique session identifier), `test_group` (numerical group assignment), and `event_name` (user interactions).
* **Readiness:** No immediate data type conversions are required, and the dataframe is ready for cleaning and validation.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving ab_test_results (1).csv to ab_test_results (1).csv


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os
os.listdir('/content/drive/MyDrive/')

['Untitled7.ipynb',
 'przykladowy_list_motywacyjny.gdoc',
 'Kopia przykladowy_list_motywacyjny.gdoc',
 'Jan_Ćwirko_CV (2).pdf',
 'IMG_20200724_112041.jpg',
 'Jan_Ćwirko_CV (2).gdoc',
 'Korwin - Mikke Janusz - Rzad rznie glupa.pdf',
 'IMG_20210318_205950.jpg',
 'Jan_wirko_CV (2).pdf',
 'Jan_wirko_CV (2).gdoc',
 '2.jpg',
 '3.jpg',
 '1.jpg',
 '4.jpg',
 '5.jpg',
 '6.jpg',
 '7.jpg',
 '8.jpg',
 '9.jpg',
 '10.jpg',
 '11.jpg',
 '12.jpg',
 '13.jpg',
 '14.jpg',
 '15.jpg',
 '16.jpg',
 '20240516_220756.jpg',
 'Dokument bez tytułu (1).gdoc',
 'Dokument bez tytułu.gdoc',
 'KALKULATOR WERSJA ALTERNATYWNA.xlsx',
 'Arkusz kalkulacyjny bez tytułu.gsheet',
 'prezentacja szefu.txt',
 'jan data.gsheet',
 'Netlix.XLSX',
 'My First Spreadsheet.gsheet',
 'adidas_usa_copy (1).csv',
 'adidas_usa_copy.csv',
 'adidas_usa (2).csv',
 'adidas_usa.csv',
 'adidas_usa (1).xlsx',
 'adidas_usa.xlsx',
 'Adidas retail data (1).gsheet',
 'Untitled spreadsheet (31).gsheet',
 'adidas_usa (17).gsheet',
 'adidas_usa (16).gshe

In [ ]:
import pandas as pd

file_name = 'ab_test_results (1).csv'
df = pd.read_csv(file_name)
print("Udało się! Plik został wczytany poprawnie.")
display(df.head())
print("\nInformacje o kolumnach:")
print(df.info())

Udało się! Plik został wczytany poprawnie.


date  ga_session_id  country   device continent         channel  \
0  2020-11-05     8412465707   Serbia  desktop    Europe     Paid Search   
1  2020-11-10      723393224  Georgia   mobile      Asia          Direct   
2  2020-11-15     5759054898   Israel   mobile      Asia  Organic Search   
3  2020-11-28     9547818693  Denmark  desktop    Europe  Organic Search   
4  2020-11-23     7535418017  Romania  desktop    Europe  Organic Search   

   test  test_group           event_name  value  
0     2           2  session with orders      1  
1     2           2  session with orders      1  
2     2           2  session with orders      1  
3     2           2  session with orders      1  
4     2           1  session with orders      1


Informacje o kolumnach:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214136 entries, 0 to 3214135
Data columns (total 10 columns):
 #   Column         Dtype 
---  ------         ----- 
 0   date           object
 1   ga_session_id  int64 
 2   country        object
 3   device         object
 4   continent      object
 5   channel        object
 6   test           int64 
 7   test_group     int64 
 8   event_name     object
 9   value          int64 
dtypes: int64(4), object(6)
memory usage: 245.2+ MB
None


In [ ]:
df = pd.read_csv('ab_test_results (1).csv')
print("Podgląd danych:")
display(df.head())
print("\nInformacje o typach danych:")
print(df.info())

Podgląd danych:


date  ga_session_id  country   device continent         channel  \
0  2020-11-05     8412465707   Serbia  desktop    Europe     Paid Search   
1  2020-11-10      723393224  Georgia   mobile      Asia          Direct   
2  2020-11-15     5759054898   Israel   mobile      Asia  Organic Search   
3  2020-11-28     9547818693  Denmark  desktop    Europe  Organic Search   
4  2020-11-23     7535418017  Romania  desktop    Europe  Organic Search   

   test  test_group           event_name  value  
0     2           2  session with orders      1  
1     2           2  session with orders      1  
2     2           2  session with orders      1  
3     2           2  session with orders      1  
4     2           1  session with orders      1


Informacje o typach danych:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214136 entries, 0 to 3214135
Data columns (total 10 columns):
 #   Column         Dtype 
---  ------         ----- 
 0   date           object
 1   ga_session_id  int64 
 2   country        object
 3   device         object
 4   continent      object
 5   channel        object
 6   test           int64 
 7   test_group     int64 
 8   event_name     object
 9   value          int64 
dtypes: int64(4), object(6)
memory usage: 245.2+ MB
None


# Stage 2: Statistical Significance Analysis (Global KPIs)

### Description
In this phase, we evaluate the global performance of the A/B test. The goal is to determine if the observed differences in user behavior between the Control and Test groups are statistically significant.

### Technical Methodology
* **Mapping:** Numerical group IDs (`1` and `2`) are mapped to descriptive labels (`control` and `test`) for better clarity.
* **Denominator:** We calculate the total number of unique sessions per group to establish a baseline.
* **Proportion Z-Test:** We apply a two-proportion Z-test to compare conversion rates. The test helps us distinguish between a genuine performance lift and random statistical noise.
* **KPIs:** The analysis focuses on four key metrics: `add_payment_info`, `add_shipping_info`, `begin_checkout`, and `new accounts`.

In [ ]:
import numpy as np
from statsmodels.stats.proportion import proportions_ztest

group_map = {1: 'control', 2: 'test'}
df['group_name'] = df['test_group'].map(group_map)
sessions_per_group = df[df['event_name'] == 'session'].groupby('group_name')['ga_session_id'].nunique()
metrics = [
    'add_payment_info',
    'add_shipping_info',
    'begin_checkout',
    'new accounts'
]

def run_stat_test(df, event_name, sessions):
    conversions = df[df['event_name'] == event_name].groupby('group_name')['ga_session_id'].nunique()

    if 'test' not in conversions or 'control' not in conversions:
        return {'Metric': event_name, 'Status': 'Missing Data'}

    counts = np.array([conversions['test'], conversions['control']])
    totals = np.array([sessions['test'], sessions['control']])

    stat, pval = proportions_ztest(counts, totals)

    return {
        'Metric': f"{event_name} / session",
        'Control CR': f"{(counts[1]/totals[1]):.2%}",
        'Test CR': f"{(counts[0]/totals[0]):.2%}",
        'P-Value': f"{pval:.4f}",
        'Is Significant?': 'YES' if pval < 0.05 else 'No'
    }

final_results = [run_stat_test(df, m, sessions_per_group) for m in metrics]
english_summary_df = pd.DataFrame(final_results)

print("A/B Test Statistical Analysis Results:")
display(english_summary_df)

A/B Test Statistical Analysis Results:


Metric Control CR Test CR P-Value Is Significant?
0   add_payment_info / session      2.02%   2.06%  0.4561              No
1  add_shipping_info / session      3.40%   3.42%  0.7272              No
2     begin_checkout / session      3.40%   3.42%  0.7335              No
3       new accounts / session      8.43%   8.29%  0.1171              No

### Analysis & Conclusions (Global Results)
* **Statistical Significance:** None of the analyzed metrics reached the significance threshold of **p < 0.05**. This means we cannot confidently state that the changes in the Test group had any impact on global conversion rates.
* **Funnel Performance:** The conversion rates for checkout-related events (payment, shipping, and initiation) remained nearly identical between both groups, with a marginal difference of only ~0.02%.
* **Account Creation:** The `new accounts` metric showed a slight decrease in the Test group (from 8.43% to 8.29%), but with a P-Value of 0.117, this result is still considered statistically insignificant.
* **Final Verdict:** At a global level, the A/B test is **neutral**. There is no evidence of a "winner," which justifies a deeper look into specific segments (countries or devices).

# Stage 3: Granular Segment Analysis (Dimensions Breakdown)

### Description
In this final analytical stage, we move beyond global averages to perform a deep dive into the data by breaking down the results into specific dimensions: `country`, `device`, `channel`, and `continent`. This approach is essential because an experiment that appears neutral at a global scale may have hidden winning or losing segments that are critical for business decisions.

### Technical Methodology
* **Dynamic Segmentation:** Instead of hard-coding specific values, the code automatically identifies all unique segments within the columns: `country`, `device`, `channel`, and `continent`.
* **Automated Looping:** We utilize a Python loop to iterate through every unique segment, creating a filtered subset of data for each iteration.
* **Reusable Function:** The `calculate_significance` function is called repeatedly for each segment, ensuring consistent calculation of the Z-test and Conversion Rates.
* **Safety Logic:** The function includes a check to skip segments that do not have enough data (e.g., sessions missing in either the Control or Test group) to prevent mathematical errors and ensure reliability.

In [ ]:
target_metrics = ['add_payment_info', 'add_shipping_info', 'begin_checkout', 'new accounts']
dimensions = ['country', 'device', 'channel', 'continent']

def calculate_significance(data, metric_list):
    """
    Calculates A/B test results for a given slice of data.
    """
    # Calculate total sessions (denominator) per group
    # Group 1 = Control, Group 2 = Test
    sessions = data[data['event_name'] == 'session'].groupby('test_group')['ga_session_id'].nunique()

    # Ensure both groups exist in the current data slice
    if 1 not in sessions or 2 not in sessions:
        return []

    results = []
    for metric in metric_list:
        # Count unique sessions with the event (numerator)
        conversions = data[data['event_name'] == metric].groupby('test_group')['ga_session_id'].nunique()

        count_control = conversions.get(1, 0)
        count_test = conversions.get(2, 0)
        total_control = sessions[1]
        total_test = sessions[2]

        # Statistical Z-test
        counts = np.array([count_test, count_control])
        totals = np.array([total_test, total_control])

        # Avoid errors if zero conversions
        if count_control == 0 or count_test == 0:
            stat, pval = (0, 1.0)
        else:
            stat, pval = proportions_ztest(counts, totals)

        results.append({
            'Metric': f"{metric} / session",
            'Control CR': count_control / total_control,
            'Test CR': count_test / total_test,
            'P-Value': pval,
            'Significant': 'YES' if pval < 0.05 else 'No'
        })
    return results

# --- EXECUTION ---

# A. Global Significance (Whole Test)
print("--- CALCULATING GLOBAL SIGNIFICANCE ---")
global_results = calculate_significance(df, target_metrics)
global_summary = pd.DataFrame(global_results)
display(global_summary)

# B. Breakdown Significance (Loops for countries, devices, etc.)
print("\n--- CALCULATING BREAKDOWNS (By Country, Device, Channel) ---")
all_breakdown_results = []

for dim in dimensions:
    unique_values = df[dim].dropna().unique()
    for val in unique_values:
        # Filter data for the specific segment (e.g., Country == 'Poland')
        subset = df[df[dim] == val]

        # Run significance for this segment
        segment_results = calculate_significance(subset, target_metrics)

        # Add metadata and store
        for res in segment_results:
            res['Dimension'] = dim
            res['Segment'] = val
            all_breakdown_results.append(res)

breakdown_summary = pd.DataFrame(all_breakdown_results)

breakdown_summary = breakdown_summary[['Dimension', 'Segment', 'Metric', 'Control CR', 'Test CR', 'P-Value', 'Significant']]
display(breakdown_summary.head(20)) # Displaying first 20 results

--- CALCULATING GLOBAL SIGNIFICANCE ---


Metric  Control CR   Test CR   P-Value Significant
0   add_payment_info / session    0.020244  0.020565  0.456084          No
1  add_shipping_info / session    0.034041  0.034234  0.727182          No
2     begin_checkout / session    0.034046  0.034234  0.733501          No
3       new accounts / session    0.084253  0.082931  0.117068          No


--- CALCULATING BREAKDOWNS (By Country, Device, Channel) ---


Dimension  Segment                       Metric  Control CR   Test CR  \
0    country   Serbia   add_payment_info / session    0.040230  0.033520   
1    country   Serbia  add_shipping_info / session    0.057471  0.039106   
2    country   Serbia     begin_checkout / session    0.057471  0.039106   
3    country   Serbia       new accounts / session    0.091954  0.089385   
4    country  Georgia   add_payment_info / session    0.053191  0.053191   
5    country  Georgia  add_shipping_info / session    0.053191  0.053191   
6    country  Georgia     begin_checkout / session    0.053191  0.053191   
7    country  Georgia       new accounts / session    0.042553  0.053191   
8    country   Israel   add_payment_info / session    0.017454  0.015355   
9    country   Israel  add_shipping_info / session    0.030801  0.026871   
10   country   Israel     begin_checkout / session    0.030801  0.026871   
11   country   Israel       new accounts / session    0.083162  0.075816   
12   country  Denmark   add_payment_info / session    0.031148  0.023217   
13   country  Denmark  add_shipping_info / session    0.040984  0.038143   
14   country  Denmark     begin_checkout / session    0.040984  0.038143   
15   country  Denmark       new accounts / session    0.078689  0.084577   
16   country  Romania   add_payment_info / session    0.005464  0.014733   
17   country  Romania  add_shipping_info / session    0.016393  0.027624   
18   country  Romania     begin_checkout / session    0.016393  0.027624   
19   country  Romania       new accounts / session    0.074681  0.090239   

     P-Value Significant  
0   0.737871          No  
1   0.420400          No  
2   0.420400          No  
3   0.933030          No  
4   1.000000          No  
5   1.000000          No  
6   1.000000          No  
7   0.732644          No  
8   0.710562          No  
9   0.597893          No  
10  0.597893          No  
11  0.541993          No  
12  0.395949          No  
13  0.799674          No  
14  0.799674          No  
15  0.707998          No  
16  0.125141          No  
17  0.205632          No  
18  0.205632          No  
19  0.349911          No

### Analysis & Conclusions (Segmented Results)

* **Consistency Across Segments:** The segmented analysis largely mirrors the global results. In most key markets such as **Serbia**, **Georgia**, and **Israel**, the changes between groups were negligible, with P-Values remaining high.
* **Interesting Observations - Romania:** Romania stands out as an interesting case. For `add_payment_info`, the Test group conversion rate (1.47%) is nearly triple that of the Control group (0.54%). However, with a P-Value of **0.12**, this does not yet meet the rigorous standards for statistical significance (p < 0.05).
* **Zero-Impact Segments:** In certain segments, like **Georgia**, the conversion rates remained identical across both groups (P-Value: 1.0), indicating that the test variation had absolutely no effect on user behavior in that region.
* **Final Summary of Python Phase:** * Globally: **No Significant Winner.**
    * Segment-wise: **No Significant Winner identified.**
    * **Next Step:** Despite the lack of statistical significance, we will now proceed to visualize these distributions in **Tableau**. Visualization often reveals trends and "near-significant" patterns that are difficult to spot in raw tables, which can be valuable for future hypothesis generation.

# Stage 4: Data Export for Visualization

### Description
The final step in our Python workflow is to export the processed statistical results. By saving the `breakdown_summary` table as a CSV file, we ensure that all complex calculations (Conversion Rates, P-Values, and Significance status) are preserved and ready for professional visualization in Tableau.

### Technical Steps
* **CSV Serialization:** Using the `.to_csv()` method to convert the DataFrame into a portable file format.
* **File Download:** Utilizing the `google.colab.files` module to trigger a local download of the results directly to the user's computer.

In [ ]:
breakdown_summary.to_csv('ab_test_final_results.csv', index=False)
from google.colab import files
files.download('ab_test_final_results.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### Summary of the Python Phase
* **Analytical Readiness:** We have successfully transitioned from raw log data to a structured, statistically-verified summary table.
* **Data Portability:** The exported file `ab_test_final_results.csv` contains all the necessary dimensions and metrics for interactive dashboarding.
* **Final Note:** While the Python analysis confirmed a lack of global significance, the exported dataset is rich enough to allow for deep-dive explorations and visual storytelling in the next phase of the project.

# Visualization and Interactive Reporting in Tableau

### Project Description and Methodology
In this final phase, the statistical results derived from the Python analysis were transformed into an interactive "Business Cockpit" dashboard. This allows stakeholders to explore the A/B test performance across various dimensions and metrics dynamically.

* **Data Source:** The `ab_test_final_results.csv` file generated during the Python analysis phase.
* **Statistical Foundation:** Z-Test for Proportions with a 95% Confidence Level ($\alpha = 0.05$).
* **Tools Used:** Google Colab (Python/Statsmodels) and Tableau Desktop/Public.

### Project Links
* **Interactive Tableau Dashboard:** [https://public.tableau.com/app/profile/jan.cwirko/viz/ABTESTANALYSISPORTFOLIO/Dashboard1?publish=yes]
* **Final Results Dataset (CSV):** [https://drive.google.com/file/d/1PzOPhP-SNpZifnjUQv060AJO_N8yoUF_/view?usp=sharing]

### Project Development Stages
* **Dashboard Architecture:** Designed a unique sidebar-driven layout with a navy-blue control panel to separate user inputs from the data display.
* **Interactivity & Navigation:** Implemented "Single Value List" filters for both `Metric` and `Dimension`, enabling precise segmentation (e.g., by Acquisition Channel or Continent).
* **Visual Optimization:** Removed grid lines and row banding to create a lightweight, modern interface, focusing attention on key KPIs like `P-Value` and `Test CR`.
* **Data Alignment:** Centered all table contents and expanded the dashboard width to ensure all statistical values are readable without horizontal scrolling.

### Final Insights and Recommendations
* **Lack of Statistical Significance:** Most test variations yielded a `P-Value > 0.05`, indicating that the observed differences between the Control and Test groups are likely due to random chance rather than the implemented changes.
* **Segment Exploration:** While the global results are not significant, the dashboard highlights stable conversion rates in specific segments like `Organic Search` and `Direct`, providing a baseline for future testing.
* **Business Recommendation:** Business Recommendation: Conclude the current test and reject the hypothesis that the variant changes significantly impact conversion rates. It is recommended to iterate on the test design or explore different hypotheses based on the stable segments identified in the dashboard.